File: src/engine/workers/worldgen/OreDistributor.ts
interface OreConfig {
  readonly blockId: number;
  readonly minY: number;
  readonly maxY: number;
  readonly veinsPerChunk: number;
  readonly veinSize: number;
}

const ORE_CONFIGS: readonly OreConfig[] = [
  { blockId: 16, minY: 5, maxY: 64, veinsPerChunk: 20, veinSize: 8 },
  { blockId: 15, minY: 5, maxY: 32, veinsPerChunk: 10, veinSize: 6 },
  { blockId: 14, minY: 5, maxY: 16, veinsPerChunk: 4, veinSize: 4 },
  { blockId: 56, minY: 5, maxY: 12, veinsPerChunk: 2, veinSize: 4 },
] as const;

export class OreDistributor {
  private readonly rng: () => number;

  constructor(seed: number) {
    let s = seed ^ 0x0be5;
    this.rng = () => {
      s = (Math.imul(s, 48271) + 1) | 0;
      return (s >>> 0) / 4294967296;
    };
  }

  distribute(voxels: Uint8Array, sizeX: number, sizeY: number, sizeZ: number): void {
    for (const cfg of ORE_CONFIGS) {
      for (let vein = 0; vein < cfg.veinsPerChunk; vein++) {
        let x = Math.floor(this.rng() * sizeX);
        let y = cfg.minY + Math.floor(this.rng() * (cfg.maxY - cfg.minY));
        let z = Math.floor(this.rng() * sizeZ);

        for (let step = 0; step < cfg.veinSize; step++) {
          x += this.rng() > 0.5 ? 1 : -1;
          y += this.rng() > 0.5 ? 1 : -1;
          z += this.rng() > 0.5 ? 1 : -1;
          if (x < 0 || x >= sizeX || y < 0 || y >= sizeY || z < 0 || z >= sizeZ) continue;
          const index = (y * sizeZ + z) * sizeX + x;
          if (voxels[index] === 1) voxels[index] = cfg.blockId;
        }
      }
    }
  }
}

File: src/world/generation/OreDistributor.cpp
#include "OreDistributor.hpp"
#include "world/BlockIds.hpp"
#include <cmath>

namespace voxel {

struct OreConfig { uint8_t blockId; int32_t minY, maxY, veinsPerChunk, veinSize; };
static constexpr OreConfig ORE_CONFIGS[] = {
  {BlockId::COAL_ORE, 5, 64, 20, 8},
  {BlockId::IRON_ORE, 5, 32, 12, 6},
  {BlockId::GOLD_ORE, 5, 16,  4, 4},
  // Diamond ore not yet registered in blocks.json — skip for now
};

OreDistributor::OreDistributor(uint32_t) {}

void OreDistributor::distribute(uint8_t* voxels, int32_t sizeX, int32_t sizeY, int32_t sizeZ,
                                uint32_t chunkSeed) {
  uint32_t rngState = chunkSeed ^ 0x0be5u;
  auto rng = [&]() -> float {
    rngState = (rngState * 48271u + 1u);
    return static_cast<float>(rngState) / 4294967296.0f;
  };

  for (const auto& cfg : ORE_CONFIGS) {
    for (int32_t vein = 0; vein < cfg.veinsPerChunk; ++vein) {
      int32_t x = static_cast<int32_t>(rng() * static_cast<float>(sizeX));
      int32_t y = cfg.minY + static_cast<int32_t>(rng() * static_cast<float>(cfg.maxY - cfg.minY));
      int32_t z = static_cast<int32_t>(rng() * static_cast<float>(sizeZ));

      for (int32_t step = 0; step < cfg.veinSize; ++step) {
        x += rng() > 0.5f ? 1 : -1;
        y += rng() > 0.5f ? 1 : -1;
        z += rng() > 0.5f ? 1 : -1;
        if (x < 0 || x >= sizeX || y < 0 || y >= sizeY || z < 0 || z >= sizeZ) continue;
        int32_t idx = (y * sizeZ + z) * sizeX + x;
        // Replace stone or grass with ore
        if (voxels[idx] == BlockId::STONE || voxels[idx] == BlockId::GRASS) voxels[idx] = cfg.blockId;
      }
    }
  }
}

} // namespace voxel
